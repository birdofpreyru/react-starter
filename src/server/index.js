/**
 * Server initialization.
 */

import fs from 'fs';
import path from 'path';

import Application from 'shared';
import { factory as reducerFactory } from 'reducers';
import { redux, server as serverFactory } from '@dr.pogodin/react-utils';

const mode = process.env.BABEL_ENV;

/* TODO: A quick workaround to pass build-time webpack config at the startup
 * of production build, without depedencies on development stuff. A more
 * elegant way to achieve the same will come later. */
let webpackConfig;
if (mode === 'production') {
  webpackConfig = path.resolve(__dirname, '../../.build-webpack-config.json');
  webpackConfig = JSON.parse(fs.readFileSync(webpackConfig, 'utf-8'));
} else {
  /* eslint-disable global-require */
  webpackConfig = require('../../webpack.config')(mode);
  /* eslint-enable global-require */
}

async function beforeRender(req) {
  const store = await redux.storeFactory({
    getReducerFactory: () => reducerFactory,
    httpRequest: req,
  });
  return { store };
}

global.KEEP_BUILD_INFO = true;
serverFactory(webpackConfig, {
  Application,
  beforeRender,
  devMode: mode === 'development',
});
