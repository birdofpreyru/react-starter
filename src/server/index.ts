/**
 * Server initialization.
 */

import fs from 'fs';
import path from 'path';

import { type Request } from 'express';

import { server as serverFactory } from '@dr.pogodin/react-utils';

import Application from 'shared';

const mode = process.env.NODE_ENV;

/* TODO: A quick workaround to pass build-time webpack config at the startup
 * of production build, without depedencies on development stuff. A more
 * elegant way to achieve the same will come later. */
let webpackConfig;
if (mode === 'production') {
  webpackConfig = path.resolve(__dirname, '../../.build-webpack-config.json');
  webpackConfig = JSON.parse(fs.readFileSync(webpackConfig, 'utf-8'));
} else {
  /* eslint-disable global-require */
  webpackConfig = require('../../webpack.config');
  if ('default' in webpackConfig) webpackConfig = webpackConfig.default;
  webpackConfig = webpackConfig(mode);
  /* eslint-enable global-require */
}

async function beforeRender(req: Request) {
  return {
    initialState: {
      domain: `${req.protocol}://${req.headers.host || req.hostname}`,
    },
  };
}

declare module global {
  let KEEP_BUILD_INFO: boolean;
}

global.KEEP_BUILD_INFO = true;
serverFactory!(webpackConfig, {
  Application,
  beforeRender,
  devMode: mode === 'development',

  /* Example of adding custom routes to the server. */
  onExpressJsSetup: (server) => {
    /* This sample & test endpoint replies back with headers, query,
     * and request body, received from the caller. */
    server.use('/__api__/example', (req, res) => {
      res.json({
        method: req.method,
        headers: req.headers,
        query: req.query,
        body: req.body,
      });
    });
  },
});
