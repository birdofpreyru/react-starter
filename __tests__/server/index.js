jest.mock('shared', () => 'APPLICATION');
jest.mock('../../webpack.config', () => () => ({}));

jest.mock('@dr.pogodin/react-utils', () => {
  const TRU = jest.requireActual('@dr.pogodin/react-utils');
  return {
    ...TRU,
    server: jest.fn(),
  };
});

test('Passes basic tests', () => {
  require('server');
  const { server } = require('@dr.pogodin/react-utils');
  expect(server.mock.calls).toHaveLength(1);
  expect(server.mock.calls[0]).toMatchSnapshot();
  const { beforeRender } = server.mock.calls[0][1];
  expect(
    () => beforeRender({ headers: {} }),
  ).not.toThrow();
});
