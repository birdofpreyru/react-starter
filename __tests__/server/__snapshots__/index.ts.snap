// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Passes basic tests 1`] = `
[
  {},
  {
    "Application": "APPLICATION",
    "beforeRender": [Function],
    "devMode": false,
    "onExpressJsSetup": [Function],
  },
]
`;
